docs/specs/add-kubernetes-support.md

```diff
@@ -0,0 +1,91 @@
+# Move to Upstream Kubernetes
+
+## Introduction
+The ansible-service-broker and all of the tooling around it
+(apb-examples, catasb, adn ansible-playbook-bundles) specifically target
+OpenShift as the underlying COE (container orchestration engine) when it
+should be cluster agnostic.
+
+## Problem Description
+Issues having been coming up during testing because the broker does not target
+the latest code for either the service-catalog or the underlying cluster.
+Kubernetes is the upstream community for OpenShift and the broker should be
+consumuing the latest upstream code for development purposes.
+
+This change will improve the Broker development process by having the Broker
+constantly being developed and tested against the latest upstream code. The
+Broker's functionality won't change, but all the tooling around it will.
+
+## Split Out All Cluster logic
+The target is to make the Broker cluster agnostic. Where the code path for
+any COE can be added to the Broker.
+
+Split out all Cluster logic into ```pkg/cluster/openshift.go``` and
+```pkg/cluster/kubernetes.go```.
+
+### Upstream and Downstream Code Hierarchy
+OpenShift is Kubernetes plus a few features, API name changes, and tools.
+In the core ansible-service-broker code, ```pkg/cluster/kubernetes.go```
+should be able to account for about 80% of the cluster opertations required
+by the broker. The remaining ~20% will be held in ```pkg/cluster/openshift.go```.
+
+### Hierarchy Diagram
+This diagram represents if the current state of code were split into two
+files.
+
+**pkg/cluster/kubernetes.go**
+
+| API Resource | Uses API | Uses CLI | Different between OpenShift and Kubernetes |
+|:---:|:---:|:---:|:---:|
+| Pods | X |  |  |
+| Exec |  | X |  |
+| Login |  | X |  |
+| ServiceAccount |  | X |  |
+| RBAC ClusterRoleBindings |  | X | X |
+| RBAC ClusterRoles |  | X | X |
+| Namespace |  | X | X |
+
+**pkg/cluster/openshift.go**
+
+| API Resource | Uses API | Uses CLI | Different between OpenShift and Kubernetes |
+|:---:|:---:|:---:|:---:|
+| ClusterRoleBindings |  | X | X |
+| ClusterRoles |  | X | X |
+| Project |  | X | X |
+
+## APBs
+The APBs only target OpenShift as a cluster. The APBs will differ in two ways:
+ - Packages
+ - oc-login.sh
+
+### Packages
+The main issue with packaging is that folks using Kubernetes may only want the
+Kubernetes client. The package ```origin-clients``` installs both the OpenShift
+and Kubernetes clients.  If this the container size isn't an issue, then there
+may not be a problem using ```origin-clients```.
```

Comment:
There seems to be two parts to this. 1) the code being compiled into the broker for both 2) the branches will require specific clients i.e. openshift.go would require oc client.  We could put in code to only require the specific oc client IF openshift.go is being invoked. 

There was a blog post Shawn shared with me about plugins in go programs. I haven't read it yet but if there was a way to make these pieces more "pluggable" that could help.